syntax = "proto3";

package brain.action.v1;

import "brain/shared/v1/envelope.proto";

service LanguageModelService {
  rpc Chat(ChatRequest) returns (ChatResponse);
  rpc ChatBatch(ChatBatchRequest) returns (ChatBatchResponse);
  rpc Embed(EmbedRequest) returns (EmbedResponse);
  rpc EmbedBatch(EmbedBatchRequest) returns (EmbedBatchResponse);
  rpc Health(HealthRequest) returns (HealthResponse);
}

enum ReasoningLevel {
  REASONING_LEVEL_UNSPECIFIED = 0;
  REASONING_LEVEL_QUICK       = 1;
  REASONING_LEVEL_STANDARD    = 2;
  REASONING_LEVEL_DEEP        = 3;
}

enum EmbeddingProfile {
  EMBEDDING_PROFILE_UNSPECIFIED = 0;
  EMBEDDING_PROFILE_EMBEDDING   = 1;
}

message ChatResult {
  string text     = 1;
  string provider = 2;
  string model    = 3;
}

message EmbeddingVector {
  repeated float values = 1;
  string provider       = 2;
  string model          = 3;
}

message HealthStatus {
  bool service_ready = 1;
  bool adapter_ready = 2;
  string detail      = 3;
}

message ChatPayload {
  string prompt          = 1;
  ReasoningLevel profile = 2;
}

message ChatRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  ChatPayload payload                   = 2;
}

message ChatResponse {
  brain.shared.v1.EnvelopeMeta metadata       = 1;
  ChatResult payload                          = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message ChatBatchPayload {
  repeated string prompts   = 1;
  ReasoningLevel profile    = 2;
}

message ChatBatchRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  ChatBatchPayload payload              = 2;
}

message ChatBatchResponse {
  brain.shared.v1.EnvelopeMeta metadata       = 1;
  repeated ChatResult payload                 = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message EmbedPayload {
  string text               = 1;
  EmbeddingProfile profile  = 2;
}

message EmbedRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  EmbedPayload payload                  = 2;
}

message EmbedResponse {
  brain.shared.v1.EnvelopeMeta metadata       = 1;
  EmbeddingVector payload                     = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message EmbedBatchPayload {
  repeated string texts      = 1;
  EmbeddingProfile profile   = 2;
}

message EmbedBatchRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  EmbedBatchPayload payload             = 2;
}

message EmbedBatchResponse {
  brain.shared.v1.EnvelopeMeta metadata       = 1;
  repeated EmbeddingVector payload            = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}

message HealthPayload {}

message HealthRequest {
  brain.shared.v1.EnvelopeMeta metadata = 1;
  HealthPayload payload                 = 2;
}

message HealthResponse {
  brain.shared.v1.EnvelopeMeta metadata       = 1;
  HealthStatus payload                        = 2;
  repeated brain.shared.v1.ErrorDetail errors = 3;
}
